### PR TITLE
Track the formats of data in an SDL_AudioStream

### PR DESCRIPTION
**Pull request by 0x1F9F1**
* Track the format of data added to an audio stream, to allow easily switching the source format.
* Allow creating a stream without setting the src/dst format.
* Added SDL_SetAudioStreamSpeed to allow changing playback speed (instead of manually setting the frequency).

## Existing Issue(s)
#8151

### CODE REVIEW

**Comment by icculus** [comment]
> I'm not sure if creating a stream without setting the src/dst format is a good idea, since it can cause unexpected things like SDL_GetAudioStreamAvailable() returning 0 no matter how much data you add to it.

I would have get/put/available return errors until a format is set, honestly.

**Comment by icculus** [comment]
This was my attempt, which is not debugged fully yet, fwiw:

https://gist.github.com/icculus/3ee44fffa1bfffd6f330e674964d7e59

The approach is different, even though we both ended up with "chunks", but the major difference was I was trying to split things out so the app continued to use the existing API but behind the scenes we could put/get in any format. I'm not sure this is a worthwhile goal, though, having run this far with it.

I'm going to reread your patch a few more times, but I'm inclined to favor your work over mine atm.

**Comment by 0x1F9F1** [comment]
Ah sorry, I should've checked if you were also working on it.

But yeah basically I came to the conclusion that allowing specifying the src/dst format separately added too much unnecessary complexity.

**Comment by icculus** [comment]
> Ah sorry, I should've checked if you were also working on it.

Honestly, you've been doing great work in this area, and this patch is no exception, so I should have checked with _you_.  :)

(The next audio things I'm intending to look at are #8122 and #8123, but I don't think they'll change any AudioStream internals.)

File: test/testaudiostreamdynamicresample.c [hunk 1]
```diff
@@ -23,25 +23,324 @@
...
+
+static SDL_Texture** font_textures = NULL;
+static const int font_width = 16;
+static const int font_height = 16;
+
+static SDL_Vertex* text_vertices = NULL;
+static int* text_indices = NULL;
+static int text_reserved = 0;
+
+static void init_text_renderer()
```
Comment (madebr) [inline test/testaudiostreamdynamicresample.c:98]: The [SDL_test library has functions to draw text](https://github.com/libsdl-org/SDL/blob/main/include/SDL3/SDL_test_font.h).

Comment (0x1F9F1) [reply test/testaudiostreamdynamicresample.c]: Heh, not sure how I missed that.

File: test/testaudiostreamdynamicresample.c [hunk 2]
```diff
@@ -140,20 +397,14 @@ int main(int argc, char *argv[])
...
-    slider_fill_area = slider_area;
-    slider_fill_area.w /= 2;
-
     /* Load the SDL library */
     if (!SDLTest_CommonInit(state)) {
         SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't initialize SDL: %s\n", SDL_GetError());
         return 1;
     }
 
+    FONT_CHARACTER_SIZE = 16;
```
Comment (icculus) [inline test/testaudiostreamdynamicresample.c:406]: I don't love this being a global variable in the test utilities, but let's change this after this PR merges.

**Review by icculus — CHANGES_REQUESTED** [review]
This is great, let's just make the new function's docs a little clearer about what the parameter does and merge this.

**Review by icculus — APPROVED** [review]

**Comment by icculus** [comment]
Okay, merging this!

**Comment by slouken** [comment]
What is the stream speed, why do we want functions to get and set it, and how does it relate to frequency?

**Comment by icculus** [comment]
> What is the stream speed, why do we want functions to get and set it, and how does it relate to frequency?

Basically it lets you adjust the pitch in the time domain without changing the official sample rate of the stream. Data goes in and out at the requested sample rate, but turning this knobs effectively lets you speed up or slow down the sound coming through.

Behind the scenes, the resampler handles this, since it's just one extra multiplication at startup.

It moves test/testaudiostreamdynamicresample.c from an interesting hack to something more formal. A way to visualize it would be having a car engine sound that ramps up as the player presses the accelerator (the stream speed increases)...you feed it static audio data at whatever sample rate, the device gets it in the sample rate it wants, yet the pitch is still increasing correctly as it goes along.

**Comment by 0x1F9F1** [comment]
Heh, replied just before I was going to send mine. Might as well send it anyway:

> What is the stream speed, why do we want functions to get and set it, and how does it relate to frequency?

You basically get it for free by just increasing the source frequency, or decreasing the output frequency, but neither of those are good options for a regular user. When using SDL_SetAudioStreamFormat:
* Changing the src_spec no longer affects data which is already in the queue, so you can't change the frequency once you've called SDL_PutAudioStreamData (and that was never a good idea anyway).
* Whatever is reading from the stream can change the dst_spec at-will, so you can't use that either.

So instead, it's better to have it as a separate property. Apart from the get/set functions, all it actually does is add 1 line of code: `src_freq = (int)((float)src_freq * stream->speed)`.

**Comment by slouken** [comment]
So it doesn't change the data flow rate, but it changes the sound of the audio?

**Comment by 0x1F9F1** [comment]
It does change how fast it consumes input data, so playing at 2x speed will consume the input audio twice as fast.

**Comment by slouken** [comment]
And also increases the pitch to match, I see. Can this be clarified in the documentation?

**Comment by 0x1F9F1** [comment]
> And also increases the pitch to match, I see. Can this be clarified in the documentation?

It's more that changing the pitch is an inherent effect of changing the speed. It's actually much more difficult to change one without changing the other.

How about `Changing the playback speed also changes the perceived pitch of the audio.`?

**Comment by slime73** [comment]
Is the speed function different from a sample rate (aka sample frequency) multiplier, or the same thing? As a user, if it's the same thing I'd prefer it to be named something closer to "SetSampleRateModifier" or whatever is most idiomatic, since Speed on its own is pretty ambiguous and would need documentation just to understand what it does.

**Comment by 0x1F9F1** [comment]
Yeah it's just multiplying the same rate.

Maybe `SDL_SetAudioStreamFrequency`?

**Comment by slouken** [comment]
Silly question, why doesn't someone just set a different frequency on the stream?

**Comment by 0x1F9F1** [comment]
> Silly question, why doesn't someone just set a different frequency on the stream?

See my earlier comment https://github.com/libsdl-org/SDL/pull/8181#issuecomment-1703243879

**Comment by flibitijibibo** [comment]
XAudio2 calls this the frequency ratio, which I think is a pretty good description without tripping into "sample rate" or "pitch" territory:

https://learn.microsoft.com/en-us/windows/win32/api/xaudio2/nf-xaudio2-ixaudio2sourcevoice-setfrequencyratio

**Comment by sezero** [comment]
Getting a type redefinition error after this:
```
src/audio/SDL_audiocvt.c:34: error: redefinition of typedef ‘SDL_AudioQueue’
src/audio/SDL_sysaudio.h:160: note: previous declaration of ‘SDL_AudioQueue’ was here
make[2]: *** [CMakeFiles/SDL3-shared.dir/src/audio/SDL_audiocvt.c.o] Error 1
```

Suggest the following: (or just removing the duplicated typedef in SDL_audiocvt.c):
```diff
diff --git a/src/audio/SDL_sysaudio.h b/src/audio/SDL_sysaudio.h
index ce50835..ac2112c 100644
--- a/src/audio/SDL_sysaudio.h
+++ b/src/audio/SDL_sysaudio.h
@@ -157,7 +157,7 @@ typedef struct SDL_AudioDriver
     SDL_AtomicInt shutting_down;  // non-zero during SDL_Quit, so we known not to accept any last-minute device hotplugs.
 } SDL_AudioDriver;
 
-typedef struct SDL_AudioQueue SDL_AudioQueue;
+struct SDL_AudioQueue;
 
 struct SDL_AudioStream
 {
@@ -172,7 +172,7 @@ struct SDL_AudioStream
     SDL_AudioSpec dst_spec;
     float freq_ratio;
 
-    SDL_AudioQueue* queue;
+    struct SDL_AudioQueue* queue;
 
     SDL_bool track_changed;
     Sint64 resample_offset;
```

**Comment by slouken** [comment]
Yeah, it seems like we don't need a duplicate typedef.

**Comment by sezero** [comment]
Pushed https://github.com/libsdl-org/SDL/commit/1d8dfbb223ff24e7cabab7bacce6908bbdce329a

**Pull request merged**